bindings/c/src/lib.rs

```diff
@@ -0,0 +1,195 @@
+use core::slice;
+use css_inline::{CSSInliner, InlineError, InlineOptions, Url};
+use libc::{c_char, size_t};
+use std::borrow::Cow;
+use std::io::Write;
+use std::ptr::null;
+use std::str;
+
+/// Result of CSS inlining operations
+#[repr(C)]
+pub enum CssResult {
+    /// No error.
+    Ok,
+    /// Missing a stylesheet file.
+    MissingStylesheet,
+    /// When loading a remote stylesheet, the file is not available.
+    RemoteStylesheetNotAvailable,
+    /// Error in the IO layer.
+    IoError,
+    /// Error while parsing the CSS.
+    InternalSelectorParseError,
+    /// options pointer is null.
+    NullOptions,
+    /// Some option in CSSInlinerOptions couldn't be used correctly.
+    InvalidOptions,
+}
+
+/// Configuration options for CSS inlining process.
+#[repr(C)]
+pub struct CssInlinerOptions {
+    /// Keep "style" tags after inlining.
+    pub keep_style_tags: bool,
+    /// Keep "link" tags after inlining.
+    pub keep_link_tags: bool,
+    /// Used for loading external stylesheets via relative URLs.
+    pub base_url: *const c_char,
+    /// Whether remote stylesheets should be loaded or not.
+    pub load_remote_stylesheets: bool,
+    /// Additional CSS to inline.
+    pub extra_css: *const c_char,
+    /// Pre-allocate capacity for HTML nodes during parsing.
+    /// It can improve performance when you have an estimate of the number of nodes in your HTML document.
+    pub preallocate_node_capacity: size_t,
+}
+
+/// Searches for the end of a string
+fn seek_for_string_end(string: *const c_char) -> usize {
+    let mut end: usize = 0;
+    loop {
+        unsafe {
+            let ptr = string.add(end);
+            if *ptr == 0 {
+                return end;
+            }
+            end += 1;
+        };
+    }
+}
+
+/// @brief Inline CSS from @p input & write the result to @p output with @p options.
+/// @param options configuration for the inliner.
+/// @param input html to inline.
+/// @param output buffer to save the inlined CSS.
+/// @param output_size size of @p output in bytes.
+/// @return a CSS_RESULT enum variant regarding if the operation was a success or an error occurred
+#[allow(clippy::missing_safety_doc)]
+#[no_mangle]
+pub unsafe extern "C" fn inline_to(
+    options: *const CssInlinerOptions,
+    input: *const c_char,
+    output: *mut c_char,
+    output_size: size_t,
+) -> CssResult {
+    let options = CSSInliner::new(
+        match InlineOptions::try_from(match options.as_ref() {
+            Some(ptr) => ptr,
+            None => return CssResult::NullOptions,
+        }) {
+            Ok(inline_options) => inline_options,
+            Err(_) => return CssResult::InvalidOptions,
+        },
+    );
+    let html = str::from_utf8_unchecked(slice::from_raw_parts(
+        input as *const u8,
+        seek_for_string_end(input),
+    ));
```

Comment:
I guess, we may use [CStr](https://doc.rust-lang.org/nightly/core/ffi/struct.CStr.html) here, which should do exactly this + will do UTF-8 validation (via `CStr::to_str`). UTF-8 will give some performance penalty (I expect it to be neglectable), but we'll avoid some possible UB.

Reply:
Or, maybe provide an “unchecked” version for those who don’t need such a check

Reply:
I changed the type to CStr, but it stills needs to be in UTF-8. I think the code is not less performant because checking is not that expensive. There are even [SIMD versions](https://lemire.me/blog/2020/10/20/ridiculously-fast-unicode-utf-8-validation/).

Reply:
Good point! Thanks! :)